benchmarks/benchmarks-core/src/jmh/java/io/micrometer/benchmark/core/MeterRegistrationContentionBenchmark.java

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright 2020 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.benchmark.core;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Group;
+import org.openjdk.jmh.annotations.GroupThreads;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+@Measurement(iterations = 5, time = 1)
+@Warmup(iterations = 10, time = 1)
+@Fork(3)
+@BenchmarkMode(Mode.SampleTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Group)
+public class MeterRegistrationContentionBenchmark {
+    MeterRegistry registry = new SimpleMeterRegistry();
+    AtomicLong i = new AtomicLong();
+
+    @TearDown(Level.Iteration)
+    public void clear() {
+        i.set(0);
+        registry.clear();
+    }
+
+    @Benchmark @Group("no_contention") @GroupThreads(1)
+    public void no_contention_register_counter() {
+        register_counter();
+    }
+
+    @Benchmark @Group("mild_contention") @GroupThreads(2)
+    public void mild_contention_register_counter() {
+        register_counter();
+    }
+
+    @Benchmark @Group("high_contention") @GroupThreads(8)
```

Comment:
I would say this is common, not rare. When the metric is dynamic (e.g you have a tag added based on a request parameter) you can't keep a reference to the meter in your code and have to call register every time (which handles the dedupe).

For a nice asynchronous app using Netty, num cpus * 2 is normal, while for a tomcat app we could probably expect as many as 1000 threads

Reply:
For some time after startup, there may be a variety of new meters created based on such variables (HTTP status code, HTTP method, etc.). After a period of time there would be mostly existing meters retrieved and only an infrequent new meter, assuming the cardinality is under control as expected.